runtime/apps/spdm/spdm-lib/src/hash_op.rs

```diff
@@ -0,0 +1,259 @@
+// Licensed under the Apache-2.0 license
+
+use crate::commands::digests_rsp::SpdmDigest;
+use crate::protocol::algorithms::BaseHashAlgoType;
+use sha2::{Digest, Sha256, Sha384, Sha512};
+
+#[derive(Debug)]
+pub enum HashEngineError {
+    InvalidParam,
+    Sha256Failure,
+    Sha384Failure,
+    Sha512Failure,
+    UnsupportedHashType,
+}
+
+// Represents the interface for a cryptographic hash engine used in SPDM. It can be extended to async-trait if needed.
+pub trait HashEngine {
+    /// Computes the hash of the entire input data in a single step.
+    ///
+    /// # Arguments
+    ///
+    /// * `data` - A slice of bytes representing the input data to be hashed.
+    /// * `hash_type` - The type of hash algorithm to use (e.g., SHA-256, SHA-384).
+    /// * `digest` - A mutable reference to a `SpdmDigest` where the resulting hash will be stored.
+    ///
+    /// # Returns
+    ///
+    /// * `Ok(())` if the hash computation is successful.
+    /// * `Err(HashEngineError)` if an error occurs during hashing.
+    fn hash_all(
+        &self,
+        data: &[u8],
+        hash_type: BaseHashAlgoType,
+        digest: &mut SpdmDigest,
+    ) -> Result<(), HashEngineError>;
+
+    /// Initializes the hash engine for incremental hashing.
+    ///
+    /// # Arguments
+    ///
+    /// * `hash_type` - The type of hash algorithm to use (e.g., SHA-256, SHA-384).
+    ///
+    /// # Returns
+    ///
+    /// * `Ok(())` if the initialization is successful.
+    /// * `Err(HashEngineError)` if an error occurs during initialization.
+    fn start(&mut self, hash_type: BaseHashAlgoType) -> Result<(), HashEngineError>;
+
+    /// Updates the hash engine with a chunk of data.
+    ///
+    /// This method can be called multiple times to process data incrementally.
+    ///
+    /// # Arguments
+    ///
+    /// * `data` - A slice of bytes representing the input data to be added to the hash computation.
+    ///
+    /// # Returns
+    ///
+    /// * `Ok(())` if the update is successful.
+    /// * `Err(HashEngineError)` if an error occurs during the update.
+    fn update(&mut self, data: &[u8]) -> Result<(), HashEngineError>;
+
+    /// Finalizes the hash computation and produces the resulting hash.
+    ///
+    /// # Arguments
+    ///
+    /// * `digest` - A mutable reference to a `SpdmDigest` where the resulting hash will be stored.
+    ///
+    /// # Returns
+    ///
+    /// * `Ok(())` if the finalization is successful.
+    /// * `Err(HashEngineError)` if an error occurs during finalization.
+    fn finish(&mut self, digest: &mut SpdmDigest) -> Result<(), HashEngineError>;
+}
+
+// A concrete implementation of the `HashEngine` trait utilizing the `sha2` crate.
+// This implementation is intended for development and testing purposes only.
+// Future iterations should refactor this to integrate with a cryptographic service provider via a mailbox interface.
```

Comment:
For Sha384 and Sha512, we'll be able to use Caliptra's SHA mailbox commands. But sadly it won't support SHA256. For small messages (few KB), we could potentially add a new mailbox command just for SHA256 though, but the hardware does not support restore / resume operations for SHA256.